python/cusignal/io/reader.py

```diff
@@ -0,0 +1,102 @@
+# Copyright (c) 2019-2020, NVIDIA CORPORATION.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import cupy as cp
+import numpy as np
+
+from mmap import mmap, MAP_PRIVATE, PROT_READ
+
+from ._reader_cuda import _parser
+
+
+def read_bin(file):
+    """
+    Reads binary file input GPU memory.
+
+    Parameters
+    ----------
+    file : str
+        A string of filename to be read to GPU.
+
+    Returns
+    -------
+    out : ndarray
+        An 1-dimensional array containing parsed binary data.
+
+    """
+
+    with open(file, "r+") as f:
+        mm = mmap(f.fileno(), 0, flags=MAP_PRIVATE, prot=PROT_READ,)
+        out = cp.asarray(mm)
+        mm.close()
+
+    return out
+
+
+def parse_bin(
+    in1, format, keep=True, dtype=np.complex64,
+):
+    """
+    Parse binary file
+
+    Parameters
+    ----------
+    in1 : array_like
+        The binary array to be parsed.
+    format : str
+        Dataset format specification to be used when unpacking binary.
+    keep : bool, optional
+        Option whether to delete binary data after parsing..
+    dtype : data-type, optional
+        Any object that can be interpreted as a numpy data type.
+
+    Returns
+    -------
+    out : ndarray
+        An 1-dimensional array containing parsed binary data.
+
+    """
+
+    out = _parser(in1, format, keep, dtype)
+
+    return out
+
+
+def fromfile(
+    file, format, keep=True, dtype=np.complex64,
+):
+    """
+    Read and parse binary file to GPU memory
+
+    Parameters
+    ----------
+    file : str
+        A string of filename to be read/parsed/upacked to GPU.
+    format : str
+        Dataset format specification to be used when unpacking binary.
+    keep : bool, optional
+        Option whether to delete binary data on GPU after parsing.
+    dtype : data-type, optional
+        Any object that can be interpreted as a numpy data type.
+
+    Returns
+    -------
+    out : ndarray
+        An 1-dimensional array containing parsed binary data.
+
+    """
+
+    binary = read_bin(file)
+    out = parse_bin(binary, format="sigmf", keep=keep, dtype=dtype,)
+
+    return out
```

Comment:
I like having a generic `fromfile` functionality here, as I think it'll be helpful for general tasks, but I'd rather do the following:

- [ ] Remove the format parameter from `fromfile`
- [ ] Create new function in `reader.py` called `read_sigmf`. The input here can be a filename. SigMF, AFAIK, has `<filename>.sigmf-meta` and `<filename>.sigmf-data` where `<filename>` is the same between these two files. The json read will give us more info on how to read/parse the data, from which we can call `fromfile`, `parse_bin`, etc.

SigMF can be streaming to the GPU, and I think `fromfile` may be a confusing API -- or we may need more general flexibility in the API when designing for future support.

Reply:
After some more investigating it looks like CuPy has `fromfile` https://github.com/cupy/cupy/pull/2626, but it must have slipped through documentation.

After all this, I'm not sure what benefit there is duplicating the functionality. Maybe we should target a different spec.